src/daemon/engine.go

```diff
@@ -0,0 +1,97 @@
+package daemon
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/facebook/augmented-traffic-control/src/atc_thrift"
+	"github.com/facebook/augmented-traffic-control/src/shaping"
+)
+
+type ShapingEngine struct {
+	conf   *Config
+	shaper shaping.Shaper
+	hooks  [][]Hook
+}
+
+func NewShapingEngine(conf *Config) (*ShapingEngine, error) {
+	shaper, err := shaping.GetShaper()
+	if err != nil {
+		return nil, err
+	}
+	if err := shaper.Initialize(); err != nil {
+		return nil, err
+	}
+	return buildShapingEngine(shaper, conf)
+}
+
+// Separate than NewShapingEngine so that tests have an entrypoint.
+func buildShapingEngine(shaper shaping.Shaper, conf *Config) (*ShapingEngine, error) {
+	hooks := make([][]Hook, len(all_hook_types))
+	for i := range all_hook_types {
+		hooks[i] = make([]Hook, 0, 5)
+	}
+	for _, hook := range conf.Hooks {
+		types, err := hook.GetTypes()
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range types {
+			hooks[t] = append(hooks[t], hook)
+		}
+	}
+	return &ShapingEngine{
+		conf:   conf,
+		shaper: shaper,
+		hooks:  hooks,
+	}, nil
+}
+
+func (eng *ShapingEngine) GetPlatform() atc_thrift.PlatformType {
+	return eng.shaper.GetPlatform()
+}
+
+func (eng *ShapingEngine) CreateGroup(id int64, member net.IP) error {
+	if err := eng.runHooks(GROUP_JOIN, fmt.Sprintf("%d", id), member.String()); err != nil {
```

Comment:
I guess there may be legit cases where join and create would require different actions but in our case, this is enough.

One way we could make things easier is by accepting only 1 script and passing the event as first argument (the rest of the argument will depend based on the event).